IDE/ECLIPSE/DEOS/deos_malloc.c

```diff
@@ -40,16 +42,18 @@ void free_deos(void *ptr) {
     return;
 }
 
-/* The caller of this function is responsible for copying the old data into new buffer */
-
 void *realloc_deos(void *ptr, size_t size) {
     void *newptr;
 
     if (size == 0)
         return ptr;
 
     newptr = malloc_deos(size);
-    free_deos(ptr);
+
+    if (ptr != NULL && newptr != NULL) {
+        XMEMCPY((char *) newptr, (const char *) ptr, size);
```

Comment:
If ptr was smaller than newptr it's an error to overread ptr as if it were the same size.  See TrackRealloc() in memory_track.h for an example solution.

Reply:
Good point catching the corner case. FWIW, I’ve already confirmed that with the current example code there is only one XREALLOC call made from Ed26619Update() in internal.c and XMEMCPY doesn’t get called since the arguments <ptr>=0 and size=<0x8f> are passed to the realloc_deos() function.